### ISSUE TITLE
Add support for Calls to StructureType members

### ISSUE THREAD

**Issue by schreiberx** [stfc/PSyclone#2586]
It seems that Fortran functions are not supported by PsyIR.

Please see [test_a.zip](https://github.com/stfc/PSyclone/files/15292847/test_a.zip) which creates CodeBlocks for 
fparser.two.Fortran2003.Function_Reference
fparser.two.Fortran2003.Function_Reference
fparser.two.Fortran2003.Structure_Constructor

This output has been created with branch https://github.com/stfc/PSyclone/compare/2577_support_pointer_assignments

This might be depending on #2577.

**Comment by sergisiso** [comment]
@arporter for the 3rd one it's fparser that gets it wrong
```
function foo_function_returning_real()
   real :: foo_function_returning_real
end function
...
varReal = foo_function_returning_real()
```
the last line is: CodeBlock[[<class 'fparser.two.Fortran2003.Structure_Constructor'>]]
Could fparser use its symbol table to identify the right node?

**Comment by arporter** [comment]
Potentially, yes. This is probably issue https://github.com/stfc/fparser/issues/255 in fparser.

**Comment by schreiberx** [comment]
> I tried your code with #2583 PR and I get this: ![image](https://private-user-images.githubusercontent.com/1449555/330357763-066a944d-3fe6-4980-b36d-24bd2853f6b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU2NzkwMTMsIm5iZiI6MTcxNTY3ODcxMywicGF0aCI6Ii8xNDQ5NTU1LzMzMDM1Nzc2My0wNjZhOTQ0ZC0zZmU2LTQ5ODAtYjM2ZC0yNGJkMjg1M2Y2YjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTRUMDkyNTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTdmZTZlYzM0Yzk0YWNlNjNkOGIxNzFmNWNhM2IwMWQ3Mjc3NDYxODJkMWM1M2JmYTA0YTllY2IyN2Q5ZDAzZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.jbWVqzrtawUrE8GcKTQ1qYtZGIi-TSvk1NhmzpXc22Y)
> 
> Is this what do you see too?
> 
> If that's the case, the only parts not supported are the rhs of:
> 
> ```
> dummy4 => this%foo_function_returning_pointer_oo_style()
> varReal = this%foo_function_returning_real_oo_style()
> varReal = foo_function_returning_real()
> ```

Yes, these 3 red CodeBlocks are what I'm getting.

**Comment by sergisiso** [comment]
I will focus this issue to add support for the structure/derive type calls:
```
dummy4 => this%field%foo_function_returning_pointer_oo_style(arg1, arg2)
varReal = this%array_field(3)%foo_function_returning_real_oo_style(arg1, arg2)
```
which need to be converted to a PSyIR Call:
```
Call
   children[0] "routine" : StructureReference "this%array_field(3)%foo_function_returning_real_oo_style"
   children[1] "arguments[0]" Reference "arg1"
   children[2] "arguments[1]" Reference "arg2"
```
The 3rd statement from my previous comment, which we need to disambiguate between a Type constructor vs a function call with no arguments will be left for another issue.
